### PR TITLE
Mt/use secure settings

### PR DESCRIPTION
**Pull request by matriv**
Please check commit messages

### CODE REVIEW

File: sql/src/main/java/io/crate/external/S3ClientHelper.java [hunk 1]
```diff
@@ -86,28 +110,40 @@ protected AmazonS3 initClient(@Nullable String accessKey, @Nullable String secre
...
-            try {
-                accessKey = userInfoParts[0];
-                secretKey = userInfoParts[1];
-            } catch (ArrayIndexOutOfBoundsException e) {
-                // ignore
+            if (!uri.getUserInfo().isEmpty()) {
+                throw new IllegalArgumentException("S3 credentials cannot be part of the URI but instead " +
+                                                   "must be stored inside the Crate keystore");
             }
+        } else if (uri.toString().contains("@") && uri.toString().contains(":")) {
```
Comment (andreidan) [inline sql/src/main/java/io/crate/external/S3ClientHelper.java]: should we throw the same exception as above in this case too? we don't care about the URI being well formatted at this stage but only want to fail if the user attempts to provide the credentials

File: sql/src/test/java/io/crate/external/S3ClientHelperTest.java [hunk 2]
```diff
@@ -24,18 +24,31 @@
...
 
-    private final S3ClientHelper s3ClientHelper = new S3ClientHelper();
+    private S3ClientHelper s3ClientHelper;
+
+    @Before
+    public void setupS3ClientHelper() {
+        MockSecureSettings secureSettings = new MockSecureSettings();
+        secureSettings.setString(ACCESS_KEY_SETTING_NAME, "testS3AccessKey");
+        secureSettings.setString(SECRET_KEY_SETTING_NAME, "testS3SecretKey");
+        s3ClientHelper = new S3ClientHelper(Settings.builder().setSecureSettings(secureSettings).build());
```
Comment (andreidan) [inline sql/src/test/java/io/crate/external/S3ClientHelperTest.java]: shall we add a test for anonymous mode?

Comment (matriv) [reply sql/src/test/java/io/crate/external/S3ClientHelperTest.java]: @andrei fixup for this, also using new way and not the deprecated `AmazonS3Client()` construnctors.

File: CHANGES.txt [hunk 3]
```diff
@@ -15,7 +15,18 @@ the appropriate section of the docs.
 Breaking Changes
 ================
 
-- Renamed ec2 ``cloud.aws.*`` settings to ``discovery.ec2.*`` (eg. 
+- When using S3 for ``COPY FROM`` and ``COPY TO``, the credentials to access S3
```
Comment (mfussenegger) [inline CHANGES.txt]: What's the reason for this change?

Comment (matriv) [reply CHANGES.txt]: Since we introduced the crate.keystore and the user can no longer provide credentials for S3 `CREATE REPOSITORY` through the SQL command (only through the keystore) we decided to do the same for `COPY FROM/TO` to be consistent regarding our security.

File: sql/src/main/java/io/crate/external/S3ClientHelper.java (hunk 1 above)
Comment (seut) [reply sql/src/main/java/io/crate/external/S3ClientHelper.java]: I agree, its a malformatted URI and so should the exception tell one

File: sql/src/main/java/io/crate/external/S3ClientHelper.java [hunk 4]
```diff
@@ -75,39 +68,56 @@ public AWSCredentials getCredentials() {
...
+        return client();
     }
 
-    private AmazonS3 client(@Nullable String accessKey, @Nullable String secretKey) throws IOException {
+    private AmazonS3 client() throws IOException {
+        String accessKey = ACCESS_KEY_SETTING.getConcreteSetting(ACCESS_KEY_SETTING_NAME).get(settings).toString();
+        String secretKey = SECRET_KEY_SETTING.getConcreteSetting(SECRET_KEY_SETTING_NAME).get(settings).toString();
+
+        if (accessKey.isEmpty() || secretKey.isEmpty()) {
+            if (!accessKey.isEmpty()) {
```
Comment (seut) [inline sql/src/main/java/io/crate/external/S3ClientHelper.java]: I'd just throw one exception if any is empty. I don't think it adds much for the user to differ here.

**Comment by seut** [comment]
pls check test failures

File: sql/src/main/java/io/crate/external/S3ClientHelper.java (hunk 4 above)
Comment (matriv) [reply sql/src/main/java/io/crate/external/S3ClientHelper.java]: ok!

File: CHANGES.txt [hunk 5]
```diff
@@ -15,6 +15,10 @@ the appropriate section of the docs.
 Breaking Changes
 ================
 
+- ``region`` setting for ``CREATE REPOSITORY`` has been removed. The region is
```
Comment (andreidan) [inline CHANGES.txt]: "the ``region``"

Comment (andreidan) [reply CHANGES.txt]: on the next sentence I suppose you can say "it is automatically guessed ... " to avoid repeating "the region)
